### ISSUE TITLE
Document how to not support constraints depending on the actual data (not type)

### ISSUE THREAD

**Issue by rschwarz** [jump-dev/MathOptInterface.jl#608]
If a solver does not support a specific function-set pair, this is readily communicated via the `MOI.supports` method.

However, the [documentation](http://www.juliaopt.org/MathOptInterface.jl/stable/apimanual/#JuMP-mapping-1) also mentions another case that is not covered by this method:
 
> With this in mind, developers should support all the constraint types that they want to be usable from JuMP. That said, from the perspective of JuMP, solvers can safely choose to not support the following constraints:
> 
> ScalarAffineFunction in GreaterThan, LessThan, or EqualTo with a nonzero constant in the function. Constants in the affine function should instead be moved into the parameters of the corresponding sets.
 
How should this be implemented?

**Comment by leethargo** [comment]
Throwing [AddConstraintNotAllowed](https://github.com/JuliaOpt/MathOptInterface.jl/blob/master/src/constraints.jl#L35) seems to work:

```
function MOI.add_constraint(o::Optimizer, func::MOI.ScalarAffineFunction{Float64},
                            set::S) where {S <: SS}
    if func.constant != 0.0
        msg = "SCIP does not support linear constraints with a constant offset."
        throw(MOI.AddConstraintNotAllowed{MOI.ScalarAffineFunction{Float64}, S}(msg))
    end
    [...]
```
but according to gitter discussion that error should only be used for whole constraint classes.

I guess that [UnsupportedConstraint](https://github.com/JuliaOpt/MathOptInterface.jl/blob/master/src/constraints.jl#L23) is no better, then.

**Comment by joaquimg** [comment]
In [this part](https://github.com/JuliaOpt/LinQuadOptInterface.jl/blob/7f4d09bf834f1ba3e2938397dbfae5c7108fe281/src/constraints/scalaraffine.jl#L44) of LQOI we simply move the constant to the set and send a warning, no errors...

**Comment by leethargo** [comment]
Right, this is what I did initially.
But then how do you implement
```
function MOI.set(o::SCIP.Optimizer, ::MOI.ConstraintSet, ci::CI{MOI.ScalarAffineFunction{Float64},S}, set::S)
```
if you did not store the constant somewhere in the `Optimizer`?

**Comment by joaquimg** [comment]
I think I understand what you mean, in LQOI we decided to save the constraint constant.

I am in favor of the warning making it clear that the original constraint was modified and the the modification you are asking about would replace the modifies set and not the originally user defined set.

**Comment by joaquimg** [comment]
But we need a decision here and it should be made clear in the manual.

**Comment by odow** [comment]
Per the documentation: 
https://github.com/jump-dev/MathOptInterface.jl/blob/6c5fd6bac5d8bfd6c15f723e4c87bbbfdc9a21b2/src/constraints.jl#L116-L129
You should throw `AddConstraintNotAllowed`. If it's just the non-zero constant term, throw `ScalarFunctionConstantNotZero`.

This issue is pretty old, so I'm tempted to close it; it obviously hasn't been an issue in the last two years.

**Comment by odow** [comment]
Closed by #1300
